File: GraphemeSplitterTest/TestStringExtensions.cs
using System;
using System.Text;
using Xunit;
using GraphemeSplitter;
using System.Linq;

namespace GraphemeSplitterTest
{
    public class TestStringExtensions
    {
        [Fact]
        public void GetCodePoints()
        {
            var s = "aáαℵАあ亜🐭👩𩸽";
            var expected = new[] { "a", "á", "α", "ℵ", "А", "あ", "亜", "🐭", "👩", "𩸽" };

            Span<byte> utf32Bytes = Encoding.UTF32.GetBytes(s);
            var expectedCodePoints = utf32Bytes.NonPortableCast<byte, uint>();

            var actual = s.GetCodePoints().ToArray();

            Assert.Equal(expected.Length, actual.Length);
            Assert.Equal(expectedCodePoints.Length, actual.Length);

            for (int i = 0; i < expectedCodePoints.Length; i++)
            {
                Assert.Equal(expected[i], new StringSegment(s, actual[i].index, actual[i].count).ToString());
                Assert.Equal(expectedCodePoints[i], actual[i].codePoint);
            }
        }

        public void aaa()
        {
            var s = "aáαℵАあ亜🐭👩𩸽Z͑ͫ̓ͪ̂ͫ̽͏̴̙̤̞͉͚̯̞̠͍A̴̵̜̰͔ͫ͗͢L̠ͨͧͩ͘G̴̻͈͍͔̹̑͗̎̅͛́Ǫ̵̹̻̝̳͂̌̌͘!͖̬̰̙̗̿̋ͥͥ̂ͣ̐́́͜͞👩🏻‍👩🏻‍👦🏻‍👦🏻👩🏻‍👩🏻‍👦🏻‍👦🏼👩🏻‍👩🏻‍👦🏻‍👦🏽👩🏻‍👩🏻‍👦🏻‍👦🏾";
            var expected = new[] { "a", "á", "α", "ℵ", "А", "あ", "亜", "🐭", "👩", "𩸽", "Z͑ͫ̓ͪ̂ͫ̽͏̴̙̤̞͉͚̯̞̠͍", "A̴̵̜̰͔ͫ͗͢", "L̠ͨͧͩ͘", "G̴̻͈͍͔̹̑͗̎̅͛́", "Ǫ̵̹̻̝̳͂̌̌͘", "!͖̬̰̙̗̿̋ͥͥ̂ͣ̐́́͜͞", "👩🏻‍👩🏻‍👦🏻‍👦🏻", "👩🏻‍👩🏻‍👦🏻‍👦🏼", "👩🏻‍👩🏻‍👦🏻‍👦🏽", "👩🏻‍👩🏻‍👦🏻‍👦🏾" };
            GetGraphemes(s, expected);
        }

        [Fact]
        public void GetGraphemesSingleUtf16() => GetGraphemes("aáαℵАあ亜", "a", "á", "α", "ℵ", "А", "あ", "亜");

        [Fact]
        public void GetGraphemesSingleCodePoint() => GetGraphemes("🐭👩𩸽👪", "🐭", "👩", "𩸽", "👪");

        [Fact]
        public void GetGraphemesCombining() => GetGraphemes("Z͑ͫ̓ͪ̂ͫ̽͏̴̙̤̞͉͚̯̞̠͍A̴̵̜̰͔ͫ͗͢L̠ͨͧͩ͘G̴̻͈͍͔̹̑͗̎̅͛́Ǫ̵̹̻̝̳͂̌̌͘!͖̬̰̙̗̿̋ͥͥ̂ͣ̐́́͜͞", "Z͑ͫ̓ͪ̂ͫ̽͏̴̙̤̞͉͚̯̞̠͍", "A̴̵̜̰͔ͫ͗͢", "L̠ͨͧͩ͘", "G̴̻͈͍͔̹̑͗̎̅͛́", "Ǫ̵̹̻̝̳͂̌̌͘", "!͖̬̰̙̗̿̋ͥͥ̂ͣ̐́́͜͞");

        //[Fact]
        //public void GetGraphemesHangul() => GetGraphemes("", "");

        [Fact]
        public void GetGraphemesEmojiSkinTone() => GetGraphemes("👩🏻👩🏼👩🏽👩🏾👩🏿👨🏻👨🏼👨🏽👨🏾👨🏿", "👩🏻", "👩🏼", "👩🏽", "👩🏾", "👩🏿", "👨🏻", "👨🏼", "👨🏽", "👨🏾", "👨🏿");

        [Fact]
        public void GetGraphemesZwjEmoji() => GetGraphemes("👩‍👩‍👩‍👩‍👩‍👩‍👩‍👩‍👩‍👩👩‍👩‍👧‍👧👩‍👩‍👧👩‍👧", "👩‍👩‍👩‍👩‍👩‍👩‍👩‍👩‍👩‍👩", "👩‍👩‍👧‍👧", "👩‍👩‍👧", "👩‍👧");

        private static void GetGraphemes(string s, params string[] expected)
        {
            foreach (var x in expected)
            {
                var y = x.GetGraphemes().ToArray();
                Assert.Equal(1, y.Length);
            }

            var actual = s.GetGraphemes().ToArray();

            Assert.Equal(expected.Length, actual.Length);

            for (int i = 0; i < expected.Length; i++)
            {
                Assert.Equal(expected[i], actual[i].ToString());
            }
        }
    }
}

File: GraphemeSplitterJSTest/TestStringExtensions.mjs
// orginal code: https://github.com/taisukef/GraphemeSplitter/blob/master/GraphemeSplitterTest/TestStringExtensions.cs

import Assert from './Assert.mjs'
//import gsplit from './NullSplitter.mjs'
import gsplit from '../GraphemeSplitter/StringSplitter.mjs'

const GetGraphemes = function(s, ...expected) {
    for (const x of expected) {
        const y = gsplit.split(x)
        Assert.Single(y)
    }
    const actual = gsplit.split(s)
    Assert.Equal(expected.length, actual.length)
    for (let i = 0; i < expected.length; i++) {
        Assert.Equal(expected[i], actual[i])
    }
}

const test = function() {
    // GetGraphemesSingleUtf16()
    GetGraphemes("aáαℵАあ亜", "a", "á", "α", "ℵ", "А", "あ", "亜")
    
    //GetGraphemesSingleCodePoint
    GetGraphemes("🐭👩𩸽👪", "🐭", "👩", "𩸽", "👪");

    // GetGraphemesCombining()


    GetGraphemes("Z͑ͫ̓ͪ̂ͫ̽͏̴̙̤̞͉͚̯̞̠͍A̴̵̜̰͔ͫ͗͢L̠ͨͧͩ͘G̴̻͈͍͔̹̑͗̎̅͛́Ǫ̵̹̻̝̳͂̌̌͘!͖̬̰̙̗̿̋ͥͥ̂ͣ̐́́͜͞", "Z͑ͫ̓ͪ̂ͫ̽͏̴̙̤̞͉͚̯̞̠͍", "A̴̵̜̰͔ͫ͗͢", "L̠ͨͧͩ͘", "G̴̻͈͍͔̹̑͗̎̅͛́", "Ǫ̵̹̻̝̳͂̌̌͘", "!͖̬̰̙̗̿̋ͥͥ̂ͣ̐́́͜͞");


    // GetGraphemesEmojiSkinTone()
    GetGraphemes("👩🏻👩🏼👩🏽👩🏾👩🏿👨🏻👨🏼👨🏽👨🏾👨🏿", "👩🏻", "👩🏼", "👩🏽", "👩🏾", "👩🏿", "👨🏻", "👨🏼", "👨🏽", "👨🏾", "👨🏿");

    // GetGraphemesZwjEmoji()
    GetGraphemes("👩‍👩‍👩‍👩‍👩‍👩‍👩‍👩‍👩‍👩👩‍👩‍👧‍👧👩‍👩‍👧👩‍👧", "👩‍👩‍👩‍👩‍👩‍👩‍👩‍👩‍👩‍👩", "👩‍👩‍👧‍👧", "👩‍👩‍👧", "👩‍👧");

    // GetGraphemesZwjEmojiSkinTone()
    GetGraphemes("👨🏽‍👨🏿‍👨🏿‍👩🏿‍👩🏾👩🏼‍👨🏼‍👨🏾‍👩🏿‍👩🏾👨🏽‍👩🏽‍👩🏾‍👩🏻‍👨🏿👨‍👨‍👧‍👦👩‍👩‍👧‍👦👨‍👨‍👧‍👦", "👨🏽‍👨🏿‍👨🏿‍👩🏿‍👩🏾", "👩🏼‍👨🏼‍👨🏾‍👩🏿‍👩🏾", "👨🏽‍👩🏽‍👩🏾‍👩🏻‍👨🏿", "👨‍👨‍👧‍👦", "👩‍👩‍👧‍👦", "👨‍👨‍👧‍👦");

    // GetGraphemesVariationSelector()
    GetGraphemes("吉󠄀𠮟󠄀葛葛󠄀葛󠄁", "吉󠄀", "𠮟󠄀", "葛", "葛󠄀", "葛󠄁");

    // GetGraphemesHangul()
    GetGraphemes("안녕하세요", "안", "녕", "하", "세", "요");

    // GetGraphemesHindi()
    //GetGraphemes("नमस्ते", "न", "म", "स्ते"); // fail. स्ते → स्, ते. need help

    // GetGraphemesFlagSequence()
    //GetGraphemes("", "");
}

test()
console.log("test ok!")
